### ISSUE TITLE
Anonymous subclass of ActiveRecord::Base causes exception

### ISSUE THREAD

**Issue by owst** [gregnavis/active_record_doctor#90]
As part of our test suite we run ActiveRecordDoctor and also create anonymous subclasses of `ActiveRecord::Base` - the latter causes an exception in the former

```
Traceback (most recent call last):
	9: from repro.rb:34:in `<main>'
	8: from /Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/runner.rb:14:in `run_one'
	7: from /Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/errors.rb:5:in `handle_exception'
	6: from /Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/runner.rb:15:in `block in run_one'
	5: from /Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/detectors/base.rb:11:in `run'
	4: from /Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/detectors/base.rb:40:in `run'
	3: from /Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/detectors/missing_unique_indexes.rb:32:in `detect'
	2: from /Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/detectors/base.rb:145:in `models'
	1: from /Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/detectors/base.rb:145:in `reject'
/Users/owenstephens/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/active_record_doctor-1.9.0/lib/active_record_doctor/detectors/base.rb:146:in `block in models': undefined method `start_with?' for nil:NilClass (NoMethodError)
```

here's a simple reproduction script:
```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", "= 5.2.6.2"
  gem "active_record_doctor", "= 1.9.0"
end

require "active_record"
require "active_record_doctor"

some_anonymous_model = Class.new(ActiveRecord::Base)

runner = ActiveRecordDoctor::Runner.new(ActiveRecordDoctor.load_config_with_defaults(nil))
runner.run_one(:missing_unique_indexes)
```

**Comment by fatkodima** [comment]
Why not to use
```ruby
SomeAnonymousModel = Class.new(ActiveRecord::Base)
```
?

This is not a trivial error to fix in `active_record_doctor`, since other methods used internally like `.table_exists?` won't also work.

**Comment by owst** [comment]
Hi @fatkodima,

The classes are defined in a test, and it isn't otherwise necessary to explicitly name the class (we are using RSpec and generating the subclass in a `let` call, so can refer to the anonymous class directly). As it happens, in the relevant test we explicitly assign `self.table_name` for the generated subclasses and I checked that `model.table_exists?` does work.

Is there any other reason to explicitly require `model.name` be non-nil? Perhaps `active_record_doctor` could just skip any models where `name.nil?`

**Comment by fatkodima** [comment]
Ah, ok, you're right 👍

**Comment by gregnavis** [comment]
@owst, let me know what you think about the above.

**Comment by owst** [comment]
Hi @gregnavis - Apologies for the massive amount of time that's passed on this issue without reply.

Our reason for calling `active_record_doctor` directly within our test suite is that our app has a somewhat slow startup and therefore rather than pay the startup cost twice (once for test suite loading and then again for `active_record_doctor` loading) we thought we could get away with only once. 

However, upon reflection, since it's not a supported setup and somewhat unusual, I'm happy just close this issue and pay the cost an additional time in our CI pipeline.

Thanks

**Comment by gregnavis** [comment]
Thanks for the detailed explanation, @owst. Let's keep this and the corresponding PR closed for now. active_record_doctor test suite defines classes on the fly, too, but assigns them to constants, so that they are named. I built [`transient_record`](https://github.com/gregnavis/transient_record) for this purpose and pair it with [`minitest-fork_executor`](https://github.com/gregnavis/minitest-fork_executor) for test isolation - I experienced really weird issues (including potential Active Support bugs) when testing against anonymous models, to supporting them properly may require a lot more work.
